File: core/providers/llm_router.py
"""
LLM Router - Unified interface for multiple LLM providers.

Supports:
- Grok (xAI) - Best for X/Twitter context
- Claude (Anthropic) - Best for complex reasoning
- OpenAI (GPT-4) - Widely compatible
- Ollama (Local) - Privacy-preserving, free

Features:
- Automatic fallback chain
- Task-based routing
- Cost tracking
- Rate limit handling
"""

import json
import os
import time
import urllib.request
import urllib.error
from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from typing import Optional, List, Dict, Generator
from enum import Enum


class TaskType(Enum):
    """Task types for intelligent routing."""
    ANALYSIS = "analysis"       # Analyzing data, sentiment
    GENERATION = "generation"   # Creating content
    EMBEDDING = "embedding"     # Vector embeddings
    QUICK = "quick"            # Fast, simple tasks
    CODING = "coding"          # Code generation/review


@dataclass
class LLMResponse:
    """Standardized response from any LLM."""
    content: str
    provider: str
    model: str
    input_tokens: int = 0
    output_tokens: int = 0
    latency_ms: int = 0
    cost_usd: float = 0.0
    raw_response: dict = None


class LLMProvider(ABC):
    """Abstract base class for LLM providers."""

    name: str = "base"
    models: List[str] = []

    @abstractmethod
    def chat(self, messages: List[Dict], model: str = None,
             temperature: float = 0.7, max_tokens: int = 4096) -> LLMResponse:
        """Send chat completion request."""
        pass

    @abstractmethod
    def is_available(self) -> bool:
        """Check if provider is available (has API key, etc)."""
        pass

    def stream(self, messages: List[Dict], model: str = None,
               temperature: float = 0.7) -> Generator[str, None, None]:
        """Stream response (optional, defaults to non-streaming)."""
        response = self.chat(messages, model, temperature)
        yield response.content


class GrokProvider(LLMProvider):
    """Grok (xAI) provider."""

    name = "grok"
    models = ["grok-2-latest", "grok-2-mini"]

    # Pricing per 1M tokens (as of 2024)
    pricing = {
        "grok-2-latest": {"input": 2.0, "output": 10.0},
        "grok-2-mini": {"input": 0.2, "output": 1.0}
    }

    def __init__(self, api_key: str = None):
        self.api_key = api_key or os.environ.get('XAI_API_KEY')
        self.base_url = "https://api.x.ai/v1"

    def is_available(self) -> bool:
        return bool(self.api_key)

    def chat(self, messages: List[Dict], model: str = None,
             temperature: float = 0.7, max_tokens: int = 4096) -> LLMResponse:

        model = model or "grok-2-latest"
        start_time = time.time()

        payload = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens
        }

        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}"
        }

        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            f"{self.base_url}/chat/completions",
            data=data, headers=headers, method='POST'
        )

        try:
            with urllib.request.urlopen(req, timeout=120) as response:
                result = json.loads(response.read().decode('utf-8'))

            latency = int((time.time() - start_time) * 1000)
            usage = result.get("usage", {})
            input_tokens = usage.get("prompt_tokens", 0)
            output_tokens = usage.get("completion_tokens", 0)

            # Calculate cost
            pricing = self.pricing.get(model, {"input": 2.0, "output": 10.0})
            cost = (input_tokens * pricing["input"] + output_tokens * pricing["output"]) / 1_000_000

            return LLMResponse(
                content=result["choices"][0]["message"]["content"],
                provider=self.name,
                model=model,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                latency_ms=latency,
                cost_usd=cost,
                raw_response=result
            )

        except Exception as e:
            return LLMResponse(
                content=f"[Grok Error: {e}]",
                provider=self.name,
                model=model,
                latency_ms=int((time.time() - start_time) * 1000)
            )


class ClaudeProvider(LLMProvider):
    """Claude (Anthropic) provider."""

    name = "claude"
    models = ["claude-sonnet-4-20250514", "claude-3-5-haiku-20241022", "claude-opus-4-20250514"]

    pricing = {
        "claude-sonnet-4-20250514": {"input": 3.0, "output": 15.0},
        "claude-3-5-haiku-20241022": {"input": 0.25, "output": 1.25},
        "claude-opus-4-20250514": {"input": 15.0, "output": 75.0}
    }

    def __init__(self, api_key: str = None):
        self.api_key = api_key or os.environ.get('ANTHROPIC_API_KEY')
        self.base_url = "https://api.anthropic.com/v1"

    def is_available(self) -> bool:
        return bool(self.api_key)

    def chat(self, messages: List[Dict], model: str = None,
             temperature: float = 0.7, max_tokens: int = 4096) -> LLMResponse:

        model = model or "claude-sonnet-4-20250514"
        start_time = time.time()

        # Convert messages format (handle system message)
        system = None
        api_messages = []
        for msg in messages:
            if msg["role"] == "system":
                system = msg["content"]
            else:
                api_messages.append(msg)

        payload = {
            "model": model,
            "max_tokens": max_tokens,
            "messages": api_messages,
            "temperature": temperature
        }
        if system:
            payload["system"] = system

        headers = {
            "Content-Type": "application/json",
            "x-api-key": self.api_key,
            "anthropic-version": "2023-06-01"
        }

        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            f"{self.base_url}/messages",
            data=data, headers=headers, method='POST'
        )

        try:
            with urllib.request.urlopen(req, timeout=120) as response:
                result = json.loads(response.read().decode('utf-8'))

            latency = int((time.time() - start_time) * 1000)
            usage = result.get("usage", {})
            input_tokens = usage.get("input_tokens", 0)
            output_tokens = usage.get("output_tokens", 0)

            pricing = self.pricing.get(model, {"input": 3.0, "output": 15.0})
            cost = (input_tokens * pricing["input"] + output_tokens * pricing["output"]) / 1_000_000

            return LLMResponse(
                content=result["content"][0]["text"],
                provider=self.name,
                model=model,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                latency_ms=latency,
                cost_usd=cost,
                raw_response=result
            )

        except Exception as e:
            return LLMResponse(
                content=f"[Claude Error: {e}]",
                provider=self.name,
                model=model,
                latency_ms=int((time.time() - start_time) * 1000)
            )


class OllamaProvider(LLMProvider):
    """Ollama (local) provider."""

    name = "ollama"
    models = ["llama3.2", "mistral", "codellama", "phi3"]

    def __init__(self, host: str = None, model: str = None):
        self.host = host or os.environ.get('OLLAMA_HOST', 'http://localhost:11434')
        self.default_model = model or os.environ.get('OLLAMA_MODEL', 'llama3.2')

    def is_available(self) -> bool:
        """Check if Ollama is running."""
        try:
            req = urllib.request.Request(f"{self.host}/api/tags")
            with urllib.request.urlopen(req, timeout=2) as response:
                return response.status == 200
        except Exception:
            return False

    def chat(self, messages: List[Dict], model: str = None,
             temperature: float = 0.7, max_tokens: int = 4096) -> LLMResponse:

        model = model or self.default_model
        start_time = time.time()

        payload = {
            "model": model,
            "messages": messages,
            "stream": False,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens
            }
        }

        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            f"{self.host}/api/chat",
            data=data,
            headers={"Content-Type": "application/json"},
            method='POST'
        )

        try:
            with urllib.request.urlopen(req, timeout=300) as response:
                result = json.loads(response.read().decode('utf-8'))

            latency = int((time.time() - start_time) * 1000)

            return LLMResponse(
                content=result["message"]["content"],
                provider=self.name,
                model=model,
                input_tokens=result.get("prompt_eval_count", 0),
                output_tokens=result.get("eval_count", 0),
                latency_ms=latency,
                cost_usd=0.0,  # Local = free
                raw_response=result
            )

        except Exception as e:
            return LLMResponse(
                content=f"[Ollama Error: {e}]",
                provider=self.name,
                model=model,
                latency_ms=int((time.time() - start_time) * 1000)
            )


class LLMRouter:
    """
    Intelligent LLM router with fallback and task-based routing.

    Usage:
        router = LLMRouter()

        # Simple call (auto-selects best available)
        response = router.chat("What is the capital of France?")

        # Task-based routing
        response = router.chat("Analyze this tweet", task=TaskType.ANALYSIS)

        # Force specific provider
        response = router.chat("...", provider="ollama")
    """

    def __init__(self, providers: List[str] = None):
        """
        Initialize router with provider priority.

        Args:
            providers: Priority order, e.g. ['grok', 'claude', 'ollama']
        """
        self.provider_order = providers or ['grok', 'claude', 'ollama']

        # Initialize all providers
        self._providers = {
            'grok': GrokProvider(),
            'claude': ClaudeProvider(),
            'ollama': OllamaProvider(),
        }

        # Task-based routing preferences
        self.task_routing = {
            TaskType.ANALYSIS: ['grok', 'claude', 'ollama'],
            TaskType.GENERATION: ['claude', 'grok', 'ollama'],
            TaskType.EMBEDDING: ['ollama', 'claude'],
            TaskType.QUICK: ['ollama', 'grok', 'claude'],
            TaskType.CODING: ['claude', 'ollama', 'grok'],
        }

        # Usage stats
        self.stats = {
            'total_calls': 0,
            'total_tokens': 0,
            'total_cost': 0.0,
            'by_provider': {}
        }

    def get_available_providers(self) -> List[str]:
        """Get list of currently available providers."""
        return [name for name, p in self._providers.items() if p.is_available()]

    def chat(self, prompt: str, system: str = None,
             task: TaskType = None, provider: str = None,
             temperature: float = 0.7, max_tokens: int = 4096) -> LLMResponse:
        """
        Send a chat request with automatic routing and fallback.

        Args:
            prompt: User message
            system: Optional system prompt
            task: Task type for intelligent routing
            provider: Force specific provider (skip routing)
            temperature: Creativity (0-2)
            max_tokens: Max response length

        Returns:
            LLMResponse with content and metadata
        """
        # Build messages
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        return self.chat_messages(messages, task, provider, temperature, max_tokens)

    def chat_messages(self, messages: List[Dict],
                      task: TaskType = None, provider: str = None,
                      temperature: float = 0.7, max_tokens: int = 4096) -> LLMResponse:
        """
        Send chat with full message list.
        """
        # Determine provider order
        if provider:
            order = [provider]
        elif task:
            order = self.task_routing.get(task, self.provider_order)
        else:
            order = self.provider_order

        # Try providers in order
        last_error = None
        for provider_name in order:
            p = self._providers.get(provider_name)
            if not p or not p.is_available():
                continue

            response = p.chat(messages, temperature=temperature, max_tokens=max_tokens)

            # Check for error
            if response.content.startswith('[') and 'Error' in response.content:
                last_error = response
                continue

            # Success - update stats
            self._update_stats(response)
            return response

        # All providers failed
        if last_error:
            return last_error

        return LLMResponse(
            content="[Error: No LLM providers available]",
            provider="none",
            model="none"
        )

    def _update_stats(self, response: LLMResponse):
        """Update usage statistics."""
        self.stats['total_calls'] += 1
        self.stats['total_tokens'] += response.input_tokens + response.output_tokens
        self.stats['total_cost'] += response.cost_usd

        provider = response.provider
        if provider not in self.stats['by_provider']:
            self.stats['by_provider'][provider] = {
                'calls': 0, 'tokens': 0, 'cost': 0.0
            }
        self.stats['by_provider'][provider]['calls'] += 1
        self.stats['by_provider'][provider]['tokens'] += response.input_tokens + response.output_tokens
        self.stats['by_provider'][provider]['cost'] += response.cost_usd

    def get_stats(self) -> dict:
        """Get usage statistics."""
        return self.stats


# Singleton router instance
_router: Optional[LLMRouter] = None


def get_router() -> LLMRouter:
    """Get global router instance."""
    global _router
    if _router is None:
        _router = LLMRouter()
    return _router

File: packages/server/src/lib/llm-router.ts
/**
 * LLM Router - Unified interface for multiple LLM providers
 *
 * Supports:
 * - Gemini (Google) - Fast, good for general tasks
 * - Grok (xAI) - Best for X/Twitter context
 * - Claude (Anthropic) - Best for complex reasoning
 * - Ollama (Local) - Privacy-preserving, free
 */

import { z } from "zod";

// Types
export interface LLMMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface LLMResponse {
  content: string;
  provider: string;
  model: string;
  inputTokens: number;
  outputTokens: number;
  latencyMs: number;
  costUsd: number;
}

export type TaskType =
  | "analysis"
  | "generation"
  | "quick"
  | "coding"
  | "embedding";

export type ProviderName = "gemini" | "grok" | "claude" | "ollama";

// Provider interface
interface LLMProvider {
  name: ProviderName;
  isAvailable(): boolean;
  chat(
    messages: LLMMessage[],
    options?: { model?: string; temperature?: number; maxTokens?: number }
  ): Promise<LLMResponse>;
}

// Pricing per 1M tokens (approximate)
const PRICING: Record<string, { input: number; output: number }> = {
  // Gemini
  "gemini-2.0-flash": { input: 0.075, output: 0.3 },
  "gemini-1.5-pro": { input: 1.25, output: 5.0 },
  "gemini-1.5-flash": { input: 0.075, output: 0.3 },
  // Grok
  "grok-2-latest": { input: 2.0, output: 10.0 },
  "grok-2-mini": { input: 0.2, output: 1.0 },
  // Claude
  "claude-sonnet-4-20250514": { input: 3.0, output: 15.0 },
  "claude-3-5-haiku-20241022": { input: 0.25, output: 1.25 },
  // Ollama (local = free)
  "llama3.2": { input: 0, output: 0 },
};

function calculateCost(
  model: string,
  inputTokens: number,
  outputTokens: number
): number {
  const pricing = PRICING[model] || { input: 0, output: 0 };
  return (
    (inputTokens * pricing.input + outputTokens * pricing.output) / 1_000_000
  );
}

// ============================================================================
// Gemini Provider
// ============================================================================
class GeminiProvider implements LLMProvider {
  name: ProviderName = "gemini";
  private apiKey: string;
  private baseUrl = "https://generativelanguage.googleapis.com/v1beta";
  private defaultModel = "gemini-2.0-flash";

  constructor() {
    this.apiKey = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY || "";
  }

  isAvailable(): boolean {
    return !!this.apiKey;
  }

  async chat(
    messages: LLMMessage[],
    options: { model?: string; temperature?: number; maxTokens?: number } = {}
  ): Promise<LLMResponse> {
    const model = options.model || this.defaultModel;
    const startTime = Date.now();

    // Convert messages to Gemini format
    const systemInstruction = messages.find((m) => m.role === "system");
    const contents = messages
      .filter((m) => m.role !== "system")
      .map((m) => ({
        role: m.role === "assistant" ? "model" : "user",
        parts: [{ text: m.content }],
      }));

    const body: Record<string, unknown> = {
      contents,
      generationConfig: {
        temperature: options.temperature ?? 0.7,
        maxOutputTokens: options.maxTokens ?? 4096,
      },
    };

    if (systemInstruction) {
      body.systemInstruction = { parts: [{ text: systemInstruction.content }] };
    }

    try {
      const response = await fetch(
        `${this.baseUrl}/models/${model}:generateContent?key=${this.apiKey}`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        }
      );

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Gemini API error: ${response.status} ${error}`);
      }

      const data = await response.json();
      const latencyMs = Date.now() - startTime;

      const content =
        data.candidates?.[0]?.content?.parts?.[0]?.text || "[No response]";
      const usage = data.usageMetadata || {};
      const inputTokens = usage.promptTokenCount || 0;
      const outputTokens = usage.candidatesTokenCount || 0;

      return {
        content,
        provider: this.name,
        model,
        inputTokens,
        outputTokens,
        latencyMs,
        costUsd: calculateCost(model, inputTokens, outputTokens),
      };
    } catch (error) {
      return {
        content: `[Gemini Error: ${error}]`,
        provider: this.name,
        model,
        inputTokens: 0,
        outputTokens: 0,
        latencyMs: Date.now() - startTime,
        costUsd: 0,
      };
    }
  }
}

// ============================================================================
// Grok Provider
// ============================================================================
class GrokProvider implements LLMProvider {
  name: ProviderName = "grok";
  private apiKey: string;
  private baseUrl = "https://api.x.ai/v1";
  private defaultModel = "grok-2-latest";

  constructor() {
    this.apiKey = process.env.XAI_API_KEY || "";
  }

  isAvailable(): boolean {
    return !!this.apiKey;
  }

  async chat(
    messages: LLMMessage[],
    options: { model?: string; temperature?: number; maxTokens?: number } = {}
  ): Promise<LLMResponse> {
    const model = options.model || this.defaultModel;
    const startTime = Date.now();

    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model,
          messages,
          temperature: options.temperature ?? 0.7,
          max_tokens: options.maxTokens ?? 4096,
        }),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Grok API error: ${response.status} ${error}`);
      }

      const data = await response.json();
      const latencyMs = Date.now() - startTime;

      const content = data.choices?.[0]?.message?.content || "[No response]";
      const usage = data.usage || {};
      const inputTokens = usage.prompt_tokens || 0;
      const outputTokens = usage.completion_tokens || 0;

      return {
        content,
        provider: this.name,
        model,
        inputTokens,
        outputTokens,
        latencyMs,
        costUsd: calculateCost(model, inputTokens, outputTokens),
      };
    } catch (error) {
      return {
        content: `[Grok Error: ${error}]`,
        provider: this.name,
        model,
        inputTokens: 0,
        outputTokens: 0,
        latencyMs: Date.now() - startTime,
        costUsd: 0,
      };
    }
  }
}

// ============================================================================
// Claude Provider
// ============================================================================
class ClaudeProvider implements LLMProvider {
  name: ProviderName = "claude";
  private apiKey: string;
  private baseUrl = "https://api.anthropic.com/v1";
  private defaultModel = "claude-sonnet-4-20250514";

  constructor() {
    this.apiKey = process.env.ANTHROPIC_API_KEY || "";
  }

  isAvailable(): boolean {
    return !!this.apiKey;
  }

  async chat(
    messages: LLMMessage[],
    options: { model?: string; temperature?: number; maxTokens?: number } = {}
  ): Promise<LLMResponse> {
    const model = options.model || this.defaultModel;
    const startTime = Date.now();

    // Extract system message
    const systemMsg = messages.find((m) => m.role === "system");
    const apiMessages = messages
      .filter((m) => m.role !== "system")
      .map((m) => ({ role: m.role, content: m.content }));

    try {
      const body: Record<string, unknown> = {
        model,
        max_tokens: options.maxTokens ?? 4096,
        messages: apiMessages,
        temperature: options.temperature ?? 0.7,
      };

      if (systemMsg) {
        body.system = systemMsg.content;
      }

      const response = await fetch(`${this.baseUrl}/messages`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "x-api-key": this.apiKey,
          "anthropic-version": "2023-06-01",
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Claude API error: ${response.status} ${error}`);
      }

      const data = await response.json();
      const latencyMs = Date.now() - startTime;

      const content = data.content?.[0]?.text || "[No response]";
      const usage = data.usage || {};
      const inputTokens = usage.input_tokens || 0;
      const outputTokens = usage.output_tokens || 0;

      return {
        content,
        provider: this.name,
        model,
        inputTokens,
        outputTokens,
        latencyMs,
        costUsd: calculateCost(model, inputTokens, outputTokens),
      };
    } catch (error) {
      return {
        content: `[Claude Error: ${error}]`,
        provider: this.name,
        model,
        inputTokens: 0,
        outputTokens: 0,
        latencyMs: Date.now() - startTime,
        costUsd: 0,
      };
    }
  }
}

// ============================================================================
// Ollama Provider (Local)
// ============================================================================
class OllamaProvider implements LLMProvider {
  name: ProviderName = "ollama";
  private host: string;
  private defaultModel: string;

  constructor() {
    this.host = process.env.OLLAMA_HOST || "http://localhost:11434";
    this.defaultModel = process.env.OLLAMA_MODEL || "llama3.2";
  }

  isAvailable(): boolean {
    // We'll assume it's available if not in production
    // In production, you'd check if the host is reachable
    return process.env.NODE_ENV !== "production" || !!process.env.OLLAMA_HOST;
  }

  async chat(
    messages: LLMMessage[],
    options: { model?: string; temperature?: number; maxTokens?: number } = {}
  ): Promise<LLMResponse> {
    const model = options.model || this.defaultModel;
    const startTime = Date.now();

    try {
      const response = await fetch(`${this.host}/api/chat`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          model,
          messages,
          stream: false,
          options: {
            temperature: options.temperature ?? 0.7,
            num_predict: options.maxTokens ?? 4096,
          },
        }),
      });

      if (!response.ok) {
        throw new Error(`Ollama error: ${response.status}`);
      }

      const data = await response.json();
      const latencyMs = Date.now() - startTime;

      return {
        content: data.message?.content || "[No response]",
        provider: this.name,
        model,
        inputTokens: data.prompt_eval_count || 0,
        outputTokens: data.eval_count || 0,
        latencyMs,
        costUsd: 0, // Local = free
      };
    } catch (error) {
      return {
        content: `[Ollama Error: ${error}]`,
        provider: this.name,
        model,
        inputTokens: 0,
        outputTokens: 0,
        latencyMs: Date.now() - startTime,
        costUsd: 0,
      };
    }
  }
}

// ============================================================================
// LLM Router
// ============================================================================
export class LLMRouter {
  private providers: Map<ProviderName, LLMProvider>;
  private defaultOrder: ProviderName[] = ["gemini", "grok", "claude", "ollama"];

  // Task-based routing
  private taskRouting: Record<TaskType, ProviderName[]> = {
    analysis: ["grok", "gemini", "claude", "ollama"],
    generation: ["claude", "gemini", "grok", "ollama"],
    quick: ["gemini", "ollama", "grok", "claude"],
    coding: ["claude", "gemini", "ollama", "grok"],
    embedding: ["ollama", "gemini", "claude"],
  };

  // Usage stats
  private stats = {
    totalCalls: 0,
    totalTokens: 0,
    totalCost: 0,
    byProvider: {} as Record<string, { calls: number; tokens: number; cost: number }>,
  };

  constructor() {
    this.providers = new Map([
      ["gemini", new GeminiProvider()],
      ["grok", new GrokProvider()],
      ["claude", new ClaudeProvider()],
      ["ollama", new OllamaProvider()],
    ]);
  }

  getAvailableProviders(): ProviderName[] {
    return Array.from(this.providers.entries())
      .filter(([_, p]) => p.isAvailable())
      .map(([name]) => name);
  }

  async chat(
    prompt: string,
    options: {
      system?: string;
      task?: TaskType;
      provider?: ProviderName;
      temperature?: number;
      maxTokens?: number;
    } = {}
  ): Promise<LLMResponse> {
    const messages: LLMMessage[] = [];
    if (options.system) {
      messages.push({ role: "system", content: options.system });
    }
    messages.push({ role: "user", content: prompt });

    return this.chatMessages(messages, options);
  }

  async chatMessages(
    messages: LLMMessage[],
    options: {
      task?: TaskType;
      provider?: ProviderName;
      temperature?: number;
      maxTokens?: number;
    } = {}
  ): Promise<LLMResponse> {
    // Determine provider order
    let order: ProviderName[];
    if (options.provider) {
      order = [options.provider];
    } else if (options.task) {
      order = this.taskRouting[options.task];
    } else {
      order = this.defaultOrder;
    }

    // Try providers in order
    let lastError: LLMResponse | null = null;
    for (const providerName of order) {
      const provider = this.providers.get(providerName);
      if (!provider || !provider.isAvailable()) continue;

      const response = await provider.chat(messages, {
        temperature: options.temperature,
        maxTokens: options.maxTokens,
      });

      // Check for error
      if (response.content.startsWith("[") && response.content.includes("Error")) {
        lastError = response;
        continue;
      }

      // Success - update stats
      this.updateStats(response);
      return response;
    }

    // All providers failed
    if (lastError) return lastError;

    return {
      content: "[Error: No LLM providers available]",
      provider: "none",
      model: "none",
      inputTokens: 0,
      outputTokens: 0,
      latencyMs: 0,
      costUsd: 0,
    };
  }

  private updateStats(response: LLMResponse): void {
    this.stats.totalCalls++;
    this.stats.totalTokens += response.inputTokens + response.outputTokens;
    this.stats.totalCost += response.costUsd;

    const p = response.provider;
    if (!this.stats.byProvider[p]) {
      this.stats.byProvider[p] = { calls: 0, tokens: 0, cost: 0 };
    }
    this.stats.byProvider[p].calls++;
    this.stats.byProvider[p].tokens += response.inputTokens + response.outputTokens;
    this.stats.byProvider[p].cost += response.costUsd;
  }

  getStats() {
    return { ...this.stats };
  }
}

// Singleton instance
let router: LLMRouter | null = null;

export function getLLMRouter(): LLMRouter {
  if (!router) {
    router = new LLMRouter();
  }
  return router;
}
